### PR TITLE
remove independent txn in sequence's function

### PR DESCRIPTION
**Pull request by daviszhen**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #

## What this PR does / why we need it:

make the execution of the sequence's function in only one transaction.
the transaction can be shared among the sessions. we only use the shared transaction in the sequence currently.

in previous design, the sequence's functions (nextval,setval,curval) are ran in an independent transaction.
Base on pervious conclusion, we need them in one shared transaction.

### CODE REVIEW

**Review by ouyuanning — APPROVED** [review]

**Review by ouyuanning — APPROVED** [review]

**Comment by matrix-meow** [comment]
@daviszhen Thanks for your contributions!

**Here are review comments for file pkg/frontend/authenticate_test.go:**
The pull request aims to remove independent transactions in the sequence's function and make the execution of the sequence's function in only one transaction. The changes made in the pull request are in the file `pkg/frontend/authenticate_test.go`. The change is to add a `nil` parameter to the `NewSession` function call.

There are a few issues with this pull request:

1. The title of the pull request is not descriptive enough. It should provide more information about what the pull request does and what problem it solves.

2. The body of the pull request is also not very informative. It only mentions that the pull request aims to make the execution of the sequence's function in only one transaction, but it does not explain why this change is necessary or how it will benefit the codebase.

3. The changes made in the pull request are minimal and do not provide enough context for the reviewer to understand the purpose of the change. The pull request should include more information about why this change is necessary and how it will improve the codebase.

To optimize the changes made in the pull request, the author should provide more information about the problem they are trying to solve and how this change will benefit the codebase. Additionally, they should provide more context about the changes made in the file `pkg/frontend/authenticate_test.go` and how it relates to the overall goal of the pull request. Finally, the title of the pull request should be more descriptive to help the reviewer understand the purpose of the change.

**Here are review comments for file pkg/frontend/internal_executor.go:**
The pull request aims to make an improvement by making the execution of the sequence's function in only one transaction. The previous design had the sequence's functions (nextval, setval, curval) run in an independent transaction, but the conclusion was that they needed to be in one shared transaction. 

The changes made in the pull request are in the file `pkg/frontend/internal_executor.go`. The change is a modification of the `newCmdSession` function, where a `nil` parameter is added to the `NewSession` function call.

However, the pull request lacks some important information that would help in understanding the changes made. For instance, there is no explanation of why the previous design had the sequence's functions run in an independent transaction, and why it was concluded that they needed to be in one shared transaction. This information would help in understanding the reasoning behind the change.

Additionally, the pull request lacks information on how the change was tested and what the expected outcome of the change is. It is important to have this information to ensure that the change does not introduce any new bugs or issues.

To optimize the changes made in the pull request, the pull request author could add more information to the pull request body, explaining the reasoning behind the change and how it was tested. They could also add comments to the code changes to explain what the `nil` parameter does and why it was added.

**Here are review comments for file pkg/frontend/mysql_cmd_executor_test.go:**
The pull request aims to make the execution of the sequence's function in only one transaction. The changes made in the pull request are in the file `pkg/frontend/mysql_cmd_executor_test.go`, where the `NewSession` function is called with an additional `nil` argument. 

There are no issues with the title and body of the pull request. However, the changes made in the pull request are not related to the purpose of the pull request. The changes made in the pull request are in the test file, and they are not related to the functionality of the sequence's function. 

To optimize the changes made in the pull request, the changes should be made in the actual code file where the sequence's function is implemented. Additionally, the pull request should be updated to reflect the changes made in the correct file. 

Therefore, the pull request needs to be updated to address the actual functionality of the sequence's function.

**Here are review comments for file pkg/frontend/mysql_protocol_test.go:**
The pull request aims to make the execution of the sequence's function in only one transaction by removing the independent transaction in the sequence's function. The changes made in the pull request are in the file `pkg/frontend/mysql_protocol_test.go`. The changes are related to adding a new parameter to the `NewSession` function call. The new parameter is `nil`. 

There are no issues mentioned in the pull request. The pull request type is marked as an improvement. 

The changes made in the pull request are related to the test file `mysql_protocol_test.go`. The changes are related to adding a new parameter to the `NewSession` function call. The new parameter is `nil`. 

There are no problems with the changes made in the pull request. However, the pull request description is not clear and concise. The pull request description should provide more information about the changes made and the reason for making those changes. Additionally, the pull request should mention any potential issues that may arise from the changes made. 

To optimize the changes made in the pull request, the pull request description should provide more information about the changes made and the reason for making those changes. Additionally, the pull request should mention any potential issues that may arise from the changes made.

**Here are review comments for file pkg/frontend/query_result_test.go:**
The pull request aims to make the execution of the sequence's function in only one transaction by removing the independent transaction in the sequence's function. The changes made in the pull request are in the file `pkg/frontend/query_result_test.go`. The change is to add an extra `nil` parameter to the `NewSession` function call.

However, the pull request lacks sufficient information to understand the context and the reason for the change. The title and the body of the pull request do not provide enough information about the problem that the pull request is trying to solve. The pull request does not reference any issue, which makes it difficult to understand the context and the motivation behind the change. 

Additionally, the pull request does not provide any information about the testing that has been done to ensure that the change does not introduce any regression or break any existing functionality. 

To improve the pull request, the following suggestions can be considered:

- Provide a more descriptive title that summarizes the problem that the pull request is trying to solve.
- Provide a detailed description of the problem that the pull request is trying to solve and the motivation behind the change.
- Reference the issue that the pull request is trying to address.
- Provide information about the testing that has been done to ensure that the change does not introduce any regression or break any existing functionality.
- Consider adding comments to the code to explain the reason for the change and how it works.

**Here are review comments for file pkg/frontend/routine.go:**
The pull request aims to make changes to the `pkg/frontend/routine.go` file. The changes include adding a new field `goroutineID` to the `Routine` struct, and two new methods `updateGoroutineId()` and `getGoroutineId()`. The purpose of these changes is to make the execution of the sequence's function in only one transaction instead of running them in an independent transaction. 

However, the pull request lacks some essential information that would help the reviewer understand the changes better. For example, it does not explain how the new changes will improve the performance or functionality of the code. Additionally, the pull request does not reference any issue number, which makes it difficult to track the changes made in the codebase. 

The changes made in the pull request seem to be incomplete and do not address the main issue. The pull request only adds a new field and two new methods, but it does not show how these changes will help in executing the sequence's function in one shared transaction. Therefore, the pull request needs to be updated to include the necessary changes to achieve the intended goal. 

To optimize the changes made in the pull request, the author should provide more information about the issue and how the changes will improve the code's performance. Additionally, the author should reference the issue number to help track the changes made in the codebase. Finally, the author should ensure that the changes made in the pull request are complete and address the main issue.

**Here are review comments for file pkg/frontend/routine_manager.go:**
The pull request aims to make an improvement to the execution of the sequence's function by running it in only one transaction instead of multiple independent transactions. The changes made in the file `pkg/frontend/routine_manager.go` include adding a parameter `nil` to the `NewSession` function call and adding a call to the `updateGoroutineId` function.

However, there are some issues with this pull request that need addressing:

1. Lack of context: The pull request does not provide enough context about the problem that the changes aim to solve. It would be helpful to include more information about the issue and why the changes are necessary.

2. Incomplete information: The pull request does not mention which issue(s) it fixes. It is important to include this information to help track the changes made and ensure that the issue is resolved.

3. Lack of testing: The pull request does not mention any testing that has been done to ensure that the changes do not introduce any new issues or bugs. It is important to thoroughly test any changes made to the codebase before merging them into the main branch.

4. Code readability: The changes made to the file `pkg/frontend/routine_manager.go` could be more readable. It would be helpful to include comments explaining the purpose of the added parameter and function call.

To optimize the changes made in the pull request, the following suggestions could be considered:

1. Provide more context: Include more information about the issue and why the changes are necessary. This will help reviewers understand the problem and ensure that the changes are appropriate.

2. Complete information: Mention which issue(s) the pull request fixes. This will help track the changes made and ensure that the issue is resolved.

3. Thorough testing: Ensure that the changes are thoroughly tested to avoid introducing new issues or bugs.

4. Code readability: Improve the readability of the changes made to the file `pkg/frontend/routine_manager.go` by including comments explaining the purpose of the added parameter and function call.

**Here are review comments for file pkg/frontend/session.go:**
Title: Remove independent txn in sequence's function

The title is clear and concise, indicating that the pull request aims to remove independent transactions in sequence's function.

Body:

The body of the pull request provides the following information:

- The type of pull request is an improvement.
- The pull request aims to make the execution of the sequence's function in only one transaction.
- The previous design had the sequence's functions (nextval, setval, curval) run in an independent transaction.
- The pull request concludes that the sequence's functions need to be in one shared transaction.

The body is clear and concise, providing the necessary information about the pull request.

Changes:

The changes made in the pull request are as follows:

- In the file pkg/frontend/session.go, the NewSession function has been modified to include a sharedTxnHandler parameter.
- The NewBackgroundSession function has been modified to include a shareTxn parameter.
- The isShareTxn function has been added to the BackgroundSession struct.
- The GetShareTxnBackgroundExec function has been added to the Session struct.
- The ExecSql function has been modified to use GetShareTxnBackgroundExec instead of GetBackgroundExec.

The changes made in the pull request are clear and concise, and they aim to remove independent transactions in sequence's function.

Feedback:

1. The pull request body should include more information about why the sequence's functions need to be in one shared transaction. This will help reviewers understand the reasoning behind the changes made in the pull request.

2. The panic statements in the NewSession and NewBackgroundSession functions should be replaced with error messages. This will make it easier for developers to understand the cause of the error and fix it.

3. The isShareTxn function in the BackgroundSession struct should be renamed to IsShareTxn to follow Go naming conventions.

4. The GetShareTxnBackgroundExec function in the Session struct should be renamed to GetSharedTransactionBackgroundExec to make its purpose clearer.

5. The comments in the code should be updated to reflect the changes made in the pull request.

6. The changes made in the pull request could be optimized by removing the panic statements and replacing them with error messages. Additionally, the code could be refactored to make it more readable and maintainable.

7. The pull request should include unit tests to ensure that the changes made do not introduce new bugs or regressions.

8. The pull request should be reviewed by other developers to ensure that the changes made are consistent with the project's coding standards and best practices.

**Here are review comments for file pkg/frontend/session_test.go:**
This pull request aims to make the execution of the sequence's function in only one transaction by removing the independent transaction in the sequence's function. The changes made in the pull request are in the file `pkg/frontend/session_test.go`. 

The changes made in the file `pkg/frontend/session_test.go` are all related to the `InitTxnHandler` and `NewSession` functions. The `InitTxnHandler` function now has two additional parameters, `nil, nil`, and the `NewSession` function has an additional parameter, `nil, nil`. These changes are made to ensure that the sequence's functions are executed in one shared transaction.

However, the pull request does not provide enough information about the issue that it fixes. The title and body of the pull request do not provide any information about the issue that it fixes. The pull request should provide a clear description of the issue that it fixes.

Moreover, the changes made in the pull request are not optimized. The changes made in the file `pkg/frontend/session_test.go` are only related to the `InitTxnHandler` and `NewSession` functions. The pull request should provide more information about the changes made in other files, if any. 

In conclusion, the pull request needs to provide more information about the issue that it fixes. Additionally, the changes made in the pull request are not optimized, and the pull request should provide more information about the changes made in other files, if any.

**Here are review comments for file pkg/frontend/txn.go:**
The pull request aims to make the execution of the sequence's function in only one transaction. The changes made in the pull request include adding a new field `shareTxn` to the `TxnHandler` struct, which is set to true if the `txnCtx` and `txnOp` are not nil. The `NewTxnOperator` and `NewTxn` functions are updated to return an error if `shareTxn` is true. The `IsValidTxnOperator`, `CommitTxn`, and `RollbackTxn` functions are updated to check if `shareTxn` is true before proceeding. 

However, the pull request lacks sufficient information to fully understand the changes made. Here are some issues that need addressing:

1. The pull request title is not descriptive enough. A more descriptive title would help reviewers understand what the pull request is about without having to read the body.

2. The pull request body does not provide enough context. The body should explain why the changes are necessary and how they will benefit the codebase.

3. The changes made in the pull request are not well-documented. There are no comments or explanations for the changes made in the code, making it difficult for reviewers to understand the changes.

4. The pull request lacks tests. It is important to have tests to ensure that the changes made do not introduce new bugs or regressions.

To optimize the changes made in the pull request, the following suggestions can be made:

1. Provide a more descriptive title that accurately reflects the changes made in the pull request.

2. Add more context to the pull request body to explain why the changes are necessary and how they will benefit the codebase.

3. Add comments and explanations to the code changes to help reviewers understand the changes.

4. Add tests to ensure that the changes made do not introduce new bugs or regressions.

5. Consider breaking down the changes into smaller, more manageable chunks to make it easier for reviewers to understand and review the changes.

**Here are review comments for file pkg/frontend/util_test.go:**
The pull request removes independent transactions in the sequence's function and makes the execution of the sequence's function in only one transaction. The changes made in the pull request are in the file `pkg/frontend/util_test.go`. The changes include modifying the `NewSession` function to include an additional parameter. 

However, the pull request lacks a clear description of the problem that it solves. It would be helpful to have more information on why the previous design was problematic and how the changes made in the pull request address those issues. Additionally, the pull request does not reference any issue numbers, which makes it difficult to track the changes made. 

To optimize the changes made in the pull request, the author could provide more detailed documentation on the changes made and how they improve the codebase. Additionally, referencing issue numbers in the pull request would make it easier to track the changes made.

**Here are review comments for file pkg/sql/plan/build_ddl.go:**
The pull request aims to make the execution of the sequence's function in only one transaction. The changes made in the file `pkg/sql/plan/build_ddl.go` are adding a check for a specific error code and returning a new error if it is encountered.

However, there are a few issues with this pull request:

1. The title of the pull request is not descriptive enough. It should provide more information about the changes made in the pull request.

2. The body of the pull request is also not very informative. It should provide more details about the problem being addressed and how the changes made in the pull request solve it.

3. The changes made in the file `pkg/sql/plan/build_ddl.go` are not related to the problem being addressed in the pull request. It seems like the changes were made to handle a specific error code, but it is not clear how this is related to the main problem being addressed in the pull request.

To optimize the changes made in the pull request, the following suggestions can be made:

1. Provide a more descriptive title for the pull request that clearly explains the changes being made.

2. Add more details to the body of the pull request, explaining the problem being addressed and how the changes made in the pull request solve it.

3. Remove the changes made in the file `pkg/sql/plan/build_ddl.go` as they are not related to the main problem being addressed in the pull request. If these changes are necessary, they should be made in a separate pull request with a clear explanation of why they are needed.

**Here are review comments for file pkg/sql/plan/function/seqfunc.go:**
The pull request removes independent transactions in the sequence's function and makes the execution of the sequence's function in only one transaction. The changes made in the pull request are in the file pkg/sql/plan/function/seqfunc.go. 

There are no issues with the title and body of the pull request. However, there are some issues with the changes made in the pull request:

1. The pull request removes the NewTxn, CommitTxn, and RollbackTxn functions, which are used to create, commit, and rollback transactions. These functions are removed without any explanation or justification. It is unclear why these functions were removed and what impact this will have on the codebase.

2. The pull request replaces the use of NewTxn, CommitTxn, and RollbackTxn functions with proc.TxnOperator. However, there is no explanation or justification for this change. It is unclear why this change was made and what impact it will have on the codebase.

3. The pull request removes the context parameter from the NewTxn, CommitTxn, and RollbackTxn functions. This is a problem because the context parameter is required for creating transactions. Removing the context parameter without any explanation or justification is not acceptable.

To optimize the changes made in the pull request, the following suggestions can be made:

1. Provide an explanation or justification for removing the NewTxn, CommitTxn, and RollbackTxn functions. If these functions are no longer needed, explain why. If they are being replaced with something else, explain what that is and why it is better.

2. Provide an explanation or justification for replacing the use of NewTxn, CommitTxn, and RollbackTxn functions with proc.TxnOperator. If this change is being made for performance reasons, explain why it is faster. If it is being made for some other reason, explain what that reason is.

3. Add back the context parameter to the NewTxn, CommitTxn, and RollbackTxn functions. The context parameter is required for creating transactions, and removing it without any explanation or justification is not acceptable.

File: pkg/frontend/routine.go [hunk 1]
```diff
@@ -130,6 +133,16 @@ func (rt *Routine) getConnectionID() uint32 {
 	return rt.getProtocol().ConnectionID()
 }
 
+func (rt *Routine) updateGoroutineId() {
```
Comment (fengttt) [inline pkg/frontend/routine.go:136]: Is this for debugging/logging only?

**Review by fengttt — APPROVED** [review]
We may need to revisit how sequence is impl in general but this looks good enough for 0.8

File: pkg/frontend/routine.go (hunk 1 above)
Comment (daviszhen) [reply pkg/frontend/routine.go:136]: For analyzing the hung. The hung happens sometimes. The go trace  with the the routineid logged can decide which the go routine is hung and which sql the go routine is running.

**Review by qingxinhome — APPROVED** [review]

**Review by m-schen — APPROVED** [review]

**Pull request merged**